### PR TITLE
Add snapshotter extension 

### PR DESCRIPTION
**Pull request by trinitys7**
## Description

This PR is to add the snapshotter extension to sync wasm directory :

- Move default wasm directory from runtime dictionary into chain home folder, `ibc_08-wasm_client_data` will be subfolder of ~/.simapp
- Add IteratorCode function to iterate wasmcode that saved in chain state.
- Testing:  setup testing app with snapshot extension enable.

closes: #4275

---

Before we can merge this PR, please make sure that all the following items have been
checked off. If any of the checklist items are not applicable, please leave them but
write a little note why.

- [ ] Targeted PR against correct branch (see [CONTRIBUTING.md](https://github.com/cosmos/ibc-go/blob/main/docs/dev/pull-requests.md#pull-request-targeting)).
- [ ] Linked to Github issue with discussion and accepted design OR link to spec that describes this work.
- [ ] Code follows the [module structure standards](https://github.com/cosmos/cosmos-sdk/blob/main/docs/docs/build/building-modules/11-structure.md) and [Go style guide](../docs/dev/go-style-guide.md).
- [ ] Wrote unit and integration [tests](https://github.com/cosmos/ibc-go/blob/main/testing/README.md#ibc-testing-package).
- [ ] Updated relevant documentation (`docs/`) or specification (`x/<module>/spec/`).
- [ ] Added relevant `godoc` [comments](https://blog.golang.org/godoc-documenting-go-code).
- [ ] Provide a [commit message](https://github.com/cosmos/ibc-go/blob/main/docs/dev/pull-requests.md#commit-messages) to be used for the changelog entry in the PR description for review.
- [ ] Re-reviewed `Files changed` in the Github PR explorer.
- [ ] Review `Codecov Report` in the comment section below once CI passes.

### CODE REVIEW

**Comment by trinitys7** [comment]
hi @damiannolan, now this PR allows edit from maintainers.

**Comment by alpe** [comment]
This code looks [familiar](https://github.com/CosmWasm/wasmd/blob/v0.42.0/x/wasm/keeper/snapshotter.go)  ;-)

File: testing/simapp/app.go
Comment (DimitrisJim) [inline testing/simapp/app.go:740]: note to self: move these to 08-wasm simapp after #4532 is merged

**Comment by DimitrisJim** [comment]
full review pending on merge of #4292

File: modules/light-clients/08-wasm/keeper/snapshotter.go
Comment (colin-axner) [inline modules/light-clients/08-wasm/keeper/snapshotter.go:26]: could we add godoc for all public funcs?

File: modules/light-clients/08-wasm/keeper/snapshotter.go
Comment (colin-axner) [inline modules/light-clients/08-wasm/keeper/snapshotter.go:92]: I think we can delete if unused.

File: modules/light-clients/08-wasm/keeper/snapshotter.go
Comment (colin-axner) [inline modules/light-clients/08-wasm/keeper/snapshotter.go:83]: we should pin the code as well?

File: modules/light-clients/08-wasm/keeper/snapshotter.go
Comment (colin-axner) [inline modules/light-clients/08-wasm/keeper/snapshotter.go:1]: should we link to the wasmd snapshotter as a reference in this file?

File: modules/light-clients/08-wasm/keeper/snapshotter_test.go
Comment (colin-axner) [inline modules/light-clients/08-wasm/keeper/snapshotter_test.go:31]: duplicate contract test case? Or are we assuming a node would never be provided such a snapshot?

**Review by colin-axner — COMMENTED** [review]
Reviewed everything except the iterate code. Thanks @trinitys7 for getting this work started and @DimitrisJim for pushing it along :pray:

**Comment by damiannolan** [comment]
Hey @trinitys7, apologies that this PR has been waiting in the wings so long! Finally we can pick up the remaining work on this. I will resolve the conflicts and can contribute some commits to pushing this forward. Thank you for getting this started ❤️

File: modules/light-clients/08-wasm/keeper/snapshotter.go
Comment (damiannolan) [reply modules/light-clients/08-wasm/keeper/snapshotter.go]: addressed in https://github.com/cosmos/ibc-go/pull/4723/commits/43b71dbce89f29158a9017c62d6d2a72a84b6854

File: modules/light-clients/08-wasm/keeper/snapshotter.go
Comment (damiannolan) [reply modules/light-clients/08-wasm/keeper/snapshotter.go]: addressed in https://github.com/cosmos/ibc-go/pull/4723/commits/0765a17a9ad1d825984c2af2c2595e4884cea90f

File: modules/light-clients/08-wasm/keeper/snapshotter.go
Comment (damiannolan) [reply modules/light-clients/08-wasm/keeper/snapshotter.go:1]: Done! https://github.com/cosmos/ibc-go/pull/4723/commits/d639e2d0af4758ff3e8ec242b0109f4baf83fb0c

File: modules/light-clients/08-wasm/keeper/snapshotter_test.go
Comment (damiannolan) [reply modules/light-clients/08-wasm/keeper/snapshotter_test.go:32]: duplicate contract test would fail with the current test structure on the initial store code call in 08-wasm here: https://github.com/cosmos/ibc-go/blob/feat/wasm-clients/modules/light-clients/08-wasm/keeper/keeper.go#L103-L105

We might be able to work around it tho.

But that being said, all calls to the vm `StoreCode` are idempotent so it should just behave as if its a no-op if its ever reached.

File: modules/light-clients/08-wasm/keeper/snapshotter.go
Comment (DimitrisJim) [reply modules/light-clients/08-wasm/keeper/snapshotter.go:127]: seems like wasmd does the pinning here: https://github.com/CosmWasm/wasmd/blob/c6919e140b0267d18238b366606db23794e00035/x/wasm/keeper/snapshotter.go#L121 calls it https://github.com/CosmWasm/wasmd/blob/c6919e140b0267d18238b366606db23794e00035/x/wasm/keeper/keeper.go#L1023

instead of in `restoreV1`.

File: modules/light-clients/08-wasm/types/utils.go
Comment (DimitrisJim) [inline modules/light-clients/08-wasm/types/utils.go:56]: same upstream

File: modules/light-clients/08-wasm/keeper/snapshotter.go
Comment (DimitrisJim) [inline modules/light-clients/08-wasm/keeper/snapshotter.go:1]: same upstream, comment on pinning in `finalizeV1` is the only possible issue

File: modules/light-clients/08-wasm/keeper/snapshotter.go
Comment (colin-axner) [reply modules/light-clients/08-wasm/keeper/snapshotter.go:127]: My vote is just remove the finalization and do it in restore

Comment (damiannolan) [reply modules/light-clients/08-wasm/keeper/snapshotter.go:127]: Aye, agreed! I will push some commits later. Some errors need to be cleaned up as well 👍

File: modules/light-clients/08-wasm/types/utils.go
Comment (damiannolan) [reply modules/light-clients/08-wasm/types/utils.go:56]: same upstream? 

What do you mean? Personally I'd like to see these functions copied from `x/wasm` moved into their own `ioutil` pkg

Comment (DimitrisJim) [reply modules/light-clients/08-wasm/types/utils.go:56]: ah yea, upstream indicating wasmd here.

File: modules/light-clients/08-wasm/testing/simapp/test_helpers.go
Comment (damiannolan) [inline modules/light-clients/08-wasm/testing/simapp/test_helpers.go:1]: are we happy to keep this test helpers simapp setup? 

Or do we want to try to make this work with the existing pattern of simapp setup? cc. @colin-axner @DimitrisJim

Comment (DimitrisJim) [reply modules/light-clients/08-wasm/testing/simapp/test_helpers.go:1]: not happy, but personal take is we shouldn't block on it and rather clean it up after milestone

Comment (colin-axner) [reply modules/light-clients/08-wasm/testing/simapp/test_helpers.go:1]: Agree ^ for v0.1, it is fine. But lets open an issue so this is tracked somewhere

**Review by colin-axner — APPROVED** [review]
Amazing work! Everything looks great to me, only comment would be to add an assertion that the dest wasm app has no contracts before the snapshots are restored

File: modules/light-clients/08-wasm/testing/simapp/test_helpers.go
Comment (damiannolan) [reply modules/light-clients/08-wasm/testing/simapp/test_helpers.go:1]: Sounds good to me! I can open an issue. We may find needing to configure something extra for snapshots in e2es which will contribute towards the work anyways :)

File: modules/light-clients/08-wasm/keeper/snapshotter_test.go
Comment (damiannolan) [reply modules/light-clients/08-wasm/keeper/snapshotter_test.go:76]: Added a couple of more lines to the test code for this :)

**Review by damiannolan — APPROVED** [review]
Dropping my approval on this now, I think its looking pretty good. I'd like to try to improve the tests in future but I think this is good to move forward with for now.

Thank you for collaboration on this PR @trinitys7 @DimitrisJim,  and @colin-axner for reviews ❤️ ❤️ ❤️

**Pull request merged**
